addons/binding/org.openhab.binding.plclogo/src/main/java/org/openhab/binding/plclogo/discovery/PLCDiscoveryService.java

```diff
@@ -0,0 +1,189 @@
+
+import static org.openhab.binding.plclogo.PLCLogoBindingConstants.THING_TYPE_DEVICE;
+
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.InterfaceAddress;
+import java.net.NetworkInterface;
+import java.net.Socket;
+import java.net.SocketException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.apache.commons.net.util.SubnetUtils;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.UID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PLCDiscoveryService} is responsible for discovering devices on
+ * the current Network. It uses every Network Interface which is connected to a network.
+ *
+ * @author Alexander Falkenstern - Initial contribution
+ */
+public class PLCDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(PLCDiscoveryService.class);
+
+    // Bridge config properties
+    private static final String LOGO_HOST = "address";
+
+    private static final int DISCOVERY_TIMEOUT = 30;
+    private static final Set<ThingTypeUID> DISCOVERABLE_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_DEVICE);
+
+    private static final int CONNECTION_TIMEOUT = 500;
+    private TreeSet<String> addresses = new TreeSet<String>();
+
+    private ExecutorService executor = null;
+
+    private class Runner implements Runnable {
+        private final ReentrantLock lock = new ReentrantLock();
+        private final String host;
+
+        public Runner(final String address) {
+            Objects.requireNonNull(address, "IP may not be null");
+            this.host = address;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void run() {
+            // gets every ip which can be assigned on the given network
+            InetAddress address = null;
+            try {
+                address = InetAddress.getByName(host);
+                if (!address.isReachable(CONNECTION_TIMEOUT / 5)) {
+                    address = null;
+                }
+            } catch (IOException exception) {
+                logger.debug("LOGO! device not found at: {}.", host);
+                address = null;
+            }
+
+            final InetSocketAddress endpoint = new InetSocketAddress(host, 102);
+            if (!endpoint.isUnresolved() && (address != null)) {
```

Comment:
The benefit of exception handling is that allows you to seperate error handling from program logic but because you limit the scope of the try-catch there are some consequences:

1: you have two similar catch blocks at line 83 & 112
2: you need to reset the address at line 85 to make sure we do not end up in this block, I think the try-catch should either span the whole block or there can be two:

Some pseudo code

```java
try {
	 // get by name and check reachable

	if (//done){
		return;
	}

	try (Socket socket = new Socket();) {

	}
}
catch (IOException exception ){

}
```